jobrunner/job_executor.py

```diff
@@ -33,26 +33,95 @@ class JobDefinition:
     allow_database_access: bool  # whether this job should have access to the database
 
 
+class ExecutorState(Enum):
+    PENDING = "pending"
+    PREPARING = "preparing"
+    PREPARED = "prepared"
+    EXECUTING = "executing"
+    EXECUTED = "executed"
+    FINALIZING = "finalizing"
+    FINALIZED = "finalized"
+    GONE = "gone"
+    ERROR = "error"
+
+
+@dataclass
+class JobStatus:
+    state: ExecutorState
+    message: Optional[str]
+
+
 @dataclass
 class JobResults:
-    state: State
-    status_code: Optional[StatusCode]
-    status_message: str
-    outputs: Mapping[str, str]
+    outputs: Mapping[str, str]  # mapping of outputs to privacy levels
+    unmatched_patterns: List[str]  # list of patterns that matched no outputs
     exit_code: int
     image_id: str
 
 
 class JobAPI:
-    def run(self, job: JobDefinition) -> None:
+    """
+    API for managing job execution.
+
+    This API is called by the job-runner to manage each job it is tracking. It models the running of a job as a state
+    machine, and the methods below are the transitions between states.
+
+    They should return either:
+     - the next state if successful
+     - the current state to indicate back pressure and to retry later
+     - the ERROR state with an appropriate message if something has gone wrong.
```

Comment:
That's covered under 'something has gone wrong' I think, but we could call it out more specifically. We have to return a state, and ERROR seems appropriate?

We could also raise an exception, but we're currently not raising any, which I think has made control flow nicer.

Reply:
Hmm.

So what should it return? No other state seems to fit, do we need new one? Should it raise instead a non-transition error?

Reply:
The same state as before, I think. The transition was impossible, so the state is unchanged.

Reply:
It's almost certainly a coding error in the job-runner if this happens, so it _may_ be appropriate for the calling code to raise an exception, but the API should just shrug and get on with its day.

Reply:
So, we currently use returning the same state as before as an indication of back pressure. So we need to return something else to indicate transition error, or else some other way to indicate back pressure.

Reply:
I think we should return the same state as before if the state doesn't change. It could not change for various reasons, which can be explained in the message.

If we need to programmatically react to different reasons then we could include a "code" as well as the "message", as we have with `StatusCode` at the moment.

Reply:
Right, we do need to respond to it programmatically, and I was hoping to avoid adding an additional code, as it this case is currently the only place it's needed.

Reply:
An invalid transition attempt feels more like hard error to me.